usr/man/man/usr/sunos/man/man1/ex.1
'\" te
.\" Copyright 1989 AT&T
.\" Copyright (c) 2004, 2011, Oracle and/or its affiliates. All rights reserved.
.\" Portions Copyright (c) 1992, X/Open Company Limited All Rights Reserve
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for permission to reproduce portions of its copyrighted documentation. Original documentation from The Open Group can be obtained online at  http://www.opengroup.org/bookstore/.
.\" The Institute of Electrical and Electronics Engineers and The Open Group, have given us permission to reprint portions of their documentation. In the following statement, the phrase "this text" refers to portions of the system documentation. Portions of this text are reprinted and reproduced in electronic form in the Sun OS Reference Manual, from IEEE Std 1003.1, 2004 Edition, Standard for Information Technology -- Portable Operating System Interface (POSIX), The Open Group Base Specifications Issue 6, Copyright (C) 2001-2004 by the Institute of Electrical and Electronics Engineers, Inc and The Open Group. In the event of any discrepancy between these versions and the original IEEE and The Open Group Standard, the original IEEE and The Open Group Standard is the referee document. The original Standard can be obtained online at http://www.opengroup.org/unix/online.html.  This notice shall appear on any product containing this material.
.TH ex 1 "29 Apr 2011" "SunOS 5.11" "User Commands"
.SH NAME
ex \- text editor
.SH SYNOPSIS
.LP
.nf
\fB/usr/bin/ex\fR [\fB-|\fR \fB-s\fR] [\fB-l\fR] [\fB-L\fR] [\fB-R\fR] [\fB-r\fR [\fIfile\fR]] [\fB-t\fR \fItag\fR] 
     [\fB-v\fR] [\fB-V\fR] [\fB-w\fR\fIn\fR] [+\fIcommand\fR | \fB-c\fR \fIcommand\fR] \fIfile\fR...
.fi

.LP
.nf
\fB/usr/xpg4/bin/ex\fR [\fB-|\fR \fB-s\fR] [\fB-l\fR] [\fB-L\fR] [\fB-R\fR] [\fB-r\fR [\fIfile\fR]] 
     [\fB-t\fR \fItag\fR] [\fB-v\fR] [\fB-V\fR] [\fB-w\fR\fIn\fR]
     [+\fIcommand\fR | \fB-c\fR \fIcommand\fR] \fIfile\fR...
.fi

.LP
.nf
\fB/usr/xpg6/bin/ex\fR [\fB-|\fR \fB-s\fR] [\fB-l\fR] [\fB-L\fR] [\fB-R\fR] [\fB-r\fR [\fIfile\fR]] 
     [\fB-t\fR \fItag\fR] [\fB-v\fR] [\fB-V\fR] [\fB-w\fR\fIn\fR]
     [+\fIcommand\fR | \fB-c\fR \fIcommand\fR] \fIfile\fR...
.fi

.SH DESCRIPTION
.sp
.LP
The \fBex\fR utility is the root of a family of editors: \fBex\fR and \fBvi\fR. \fBex\fR is a superset of \fBed\fR(1), with the most notable extension being a display editing facility. Display based editing is the focus of \fBvi\fR.
.sp
.LP
If you have a \fBCRT\fR terminal, you can wish to use a display based editor; in this case see \fBvi\fR(1), which is a command which focuses on the display-editing portion of \fBex\fR.
.sp
.LP
If you have used \fBed\fR you find that, in addition to having all of the \fBed\fR commands available, \fBex\fR has a number of additional features useful on \fBCRT\fR terminals. Intelligent terminals and high speed terminals are very pleasant to use with \fBvi\fR. Generally, the \fBex\fR editor uses far more of the capabilities of terminals than \fBed\fR does, and uses the terminal capability data base (see \fBterminfo\fR(4)) and the type of the terminal you are using from the environment variable \fBTERM\fR to determine how to drive your terminal efficiently. The editor makes use of features such as insert and delete character and line in its \fBvisual\fR command (which can be abbreviated \fBvi\fR) and which is the central mode of editing when using the \fBvi\fR command.
.sp
.LP
The \fBex\fR utility contains a number of features for easily viewing the text of the file. The \fBz\fR command gives easy access to windows of text. Typing \fB^D\fR (CTRL-D) causes the editor to scroll a half-window of text and is more useful for quickly stepping through a file than just typing return. Of course, the screen-oriented \fBvisual\fR mode gives constant access to editing context.
.sp
.LP
The \fBex\fR utility gives you help when you make mistakes. The \fBundo\fR (\fBu\fR) command allows you to reverse any single change which goes astray. \fBex\fR gives you a lot of feedback, normally printing changed lines, and indicates when more than a few lines are affected by a command so that it is easy to detect when a command has affected more lines than it should have.
.sp
.LP
The editor also normally prevents overwriting existing files, unless you edited them, so that you do not accidentally overwrite a file other than the one you are editing. If the system (or editor) crashes, or you accidentally hang up the telephone, you can use the editor \fBrecover\fR command (or \fB-r\fR \fIfile\fR option) to retrieve your work. This gets you back to within a few lines of where you left off.
.sp
.LP
The \fBex\fR utility has several features for dealing with more than one file at a time. You can give it a list of files on the command line and use the \fBnext\fR (\fBn\fR) command to deal with each in turn. The \fBnext\fR command can also be given a list of file names, or a pattern as used by the shell to specify a new set of files to be dealt with. In general, file names in the editor can be formed with full shell metasyntax. The metacharacter `%' is also available in forming file names and is replaced by the name of the current file.
.sp
.LP
The editor has a group of buffers whose names are the ASCII lower-case letters (\fBa\fR-\fBz\fR). You can place text in these named buffers where it is available to be inserted elsewhere in the file. The contents of these buffers remain available when you begin editing a new file using the \fBedit\fR (\fBe\fR) command.
.sp
.LP
There is a command \fB&\fR in \fBex\fR which repeats the last \fBsubstitute\fR command. In addition, there is a confirmed substitute command. You give a range of substitutions to be done and the editor interactively asks whether each substitution is desired.
.sp
.LP
It is possible to ignore the case of letters in searches and substitutions. \fBex\fR also allows regular expressions which match words to be constructed. This is convenient, for example, in searching for the word ``edit'' if your document also contains the word ``editor.''
.sp
.LP
\fBex\fR has a set of options which you can set to tailor it to your liking. One option which is very useful is the \fBautoindent\fR option that allows the editor to supply leading white space to align text automatically. You can then use \fB^D\fR as a backtab and space or tab to move forward to align new code easily.
.sp
.LP
Miscellaneous useful features include an intelligent \fBjoin\fR (\fBj\fR) command that supplies white space between joined lines automatically, commands \fB<\fR and \fB>\fR which shift groups of lines, and the ability to filter portions of the buffer through commands such as \fBsort\fR.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.mk
.na
\fB\fB\(mi\fR | \fB-s\fR\fR
.ad
.RS 25n
.rt  
Suppresses all interactive user feedback. This is useful when processing editor scripts.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-l\fR\fR
.ad
.RS 25n
.rt  
Sets up for editing \fBLISP\fR programs.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-L\fR\fR
.ad
.RS 25n
.rt  
Lists the name of all files saved as the result of an editor or system crash.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-R\fR\fR
.ad
.RS 25n
.rt  
\fBReadonly\fR mode. The \fBreadonly\fR flag is set, preventing accidental overwriting of the file.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-r\fR \fIfile\fR\fR
.ad
.RS 25n
.rt  
Edits \fIfile\fR after an editor or system crash. (Recovers the version of \fIfile\fR that was in the buffer when the crash occurred.)
.RE

.sp
.ne 2
.mk
.na
\fB\fB-t\fR \fItag\fR\fR
.ad
.RS 25n
.rt  
Edits the file containing the \fItag\fR and positions the editor at its definition. It is an error to specify more than one \fB-t\fR option.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-v\fR\fR
.ad
.RS 25n
.rt  
Starts up in display editing state, using \fBvi\fR. You can achieve the same effect by typing the \fBvi\fR command itself.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-V\fR\fR
.ad
.RS 25n
.rt  
Verbose. When \fBex\fR commands are read by means of standard input, the input is echoed to standard error. This can be useful when processing \fBex\fR commands within shell scripts.
.RE

.sp
.ne 2
.mk
.na
\fB\fB-w\fR\fIn\fR\fR
.ad
.RS 25n
.rt  
Sets the default window size to \fIn\fR. This is useful when using the editor over a slow speed line.
.RE

.sp
.ne 2
.mk
.na
\fB\fB+\fR\fIcommand\fR | \fB-c\fR \fIcommand\fR\fR
.ad
.RS 25n
.rt  
Begins editing by executing the specified editor \fIcommand\fR (usually a search or positioning command).
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/xpg4/bin/ex\fR, \fB/usr/xpg6/bin/ex\fR\fR
.ad
.sp .6
.RS 4n
If both the \fB-t\fR \fItag\fR and the \fB-c\fR \fIcommand\fR options are given, the \fB-t\fR \fItag\fR is processed first. That is, the file containing the tag is selected by \fB-t\fR and then the command is executed.
.RE

.SH OPERANDS
.sp
.LP
The following operand is supported:
.sp
.ne 2
.mk
.na
\fB\fIfile\fR\fR
.ad
.RS 8n
.rt  
A path name of a file to be edited.
.RE

.SH USAGE
.sp
.LP
This section defines the \fBex\fR states, commands, initializing options, and scanning pattern formations.
.SS "ex States"
.sp
.ne 2
.mk
.na
\fBCommand\fR
.ad
.RS 11n
.rt  
Normal and initial state. Input prompted for by "\fB:\fR". The line kill character cancels a partial command.
.RE

.sp
.ne 2
.mk
.na
\fBInsert\fR
.ad
.RS 11n
.rt  
Entered by \fBa\fR, \fBi\fR, or \fBc\fR. Arbitrary text can be entered. Insert state normally is terminated by a line having only "\fB\&.\fR" on it, or, abnormally, with an interrupt.
.RE

.sp
.ne 2
.mk
.na
\fBVisual\fR
.ad
.RS 11n
.rt  
Entered by typing \fBvi\fR. Terminated by typing \fBQ\fR or ^\e (Control-\e).
.RE

.SS "ex Command Names and Abbreviations"
.sp
.in +2
.nf
Command    Abbrevi-   Command    Abbrevi-    Command    Abbrevi-
Name       ation      Name       ation       Name       ation

abbrev     ab         map                    set        se

append     a          mark       ma          shell      sh

args       ar         move       m           source     so

change     c          next       n           substitute s

copy       co         number     nu          unabbrev   unab

delete     d          preserve   pre         undo       u

edit       e          print      p           unmap      unm

file       f          put        pu          version    ve

global     g          quit       q           visual     vi

insert     i          read       r           write     
 w

join       j          recover    rec         xit        x

list       l          rewind     rew         yank       ya
.fi
.in -2
.sp

.SS "Join Command Arguments"
.sp
.in +2
.nf
Join     [\fIrange\fR] j[\fIoin\fR][!] [\fIcount\fR] [\fIflags\fR]
.fi
.in -2
.sp

.sp
.LP
If count is specified:
.sp
.ne 2
.mk
.na
\fB\fB/usr/bin/ex\fR, \fB/usr/xpg6/bin/ex\fR\fR
.ad
.sp .6
.RS 4n
If no address is specified, the \fBjoin\fR command behaves as if \fI2addr\fR were the current line and the current line plus \fIcount\fR (\fB\&.,. +\fR \fIcount\fR). If one address is specified, the \fBjoin\fR command behaves as if \fI2addr\fR were the specified address and the specified address plus \fIcount\fR ( \fIaddr\fR\fB,\fR \fIaddr\fR \fB+\fR \fIcount\fR). 
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/xpg4/bin/ex\fR\fR
.ad
.sp .6
.RS 4n
If no address is specified, the \fBjoin\fR command behaves as if \fI2addr\fR were the current line and the current line plus \fIcount\fR \fB-1\fR (\fB\&.,. +\fR \fIcount\fR \fB-1\fR). If one address is specified, the \fBjoin\fR command behaves as if \fI2addr\fR were the specified address and the specified address plus \fIcount\fR \fB-1\fR ( \fIaddr\fR\fB,\fR \fIaddr\fR \fB+\fR \fIcount\fR \fB-1\fR). 
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/bin/ex\fR, \fB/usr/xpg4/bin/ex\fR, \fB/usr/xpg6/bin/ex\fR\fR
.ad
.sp .6
.RS 4n
If two or more addresses are specified, the \fBjoin\fR command behaves as if an additional address, equal to the last address plus \fIcount\fR \fB-1\fR (\fIaddr1\fR\fB, ...,\fR \fIlastaddr\fR\fB,\fR \fIlastaddr\fR + \fIcount\fR \fB-1\fR), was specified.  If this results in a second address greater than the last line of the edit buffer, it is corrected to be equal to the last line of the edit buffer. 
.RE

.sp
.LP
If no count is specified: 
.sp
.ne 2
.mk
.na
\fB\fB/usr/bin/ex\fR, \fB/usr/xpg4/bin/ex\fR, \fB/usr/xpg6/bin/ex\fR\fR
.ad
.sp .6
.RS 4n

usr/src/cmd/vi/port/ex.c
/*
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
 * or http://www.opensolaris.org/os/licensing.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright (c) 1988, 2011, Oracle and/or its affiliates. All rights reserved.
 */

/*	Copyright (c) 1984, 1986, 1987, 1988, 1989 AT&T	*/
/*	  All Rights Reserved  	*/

/* Copyright (c) 1981 Regents of the University of California */


#include "ex.h"
#include "ex_argv.h"
#include "ex_temp.h"
#include "ex_tty.h"
#include <stdlib.h>
#include <locale.h>
#include <stdio.h>
#ifdef TRACE
unsigned char	tttrace[BUFSIZ];
#endif

#define	EQ(a, b)	(strcmp(a, b) == 0)

char	*strrchr();
void	init_re(void);

/*
 * The code for ex is divided as follows:
 *
 * ex.c			Entry point and routines handling interrupt, hangup
 *			signals; initialization code.
 *
 * ex_addr.c		Address parsing routines for command mode decoding.
 *			Routines to set and check address ranges on commands.
 *
 * ex_cmds.c		Command mode command decoding.
 *
 * ex_cmds2.c		Subroutines for command decoding and processing of
 *			file names in the argument list.  Routines to print
 *			messages and reset state when errors occur.
 *
 * ex_cmdsub.c		Subroutines which implement command mode functions
 *			such as append, delete, join.
 *
 * ex_data.c		Initialization of options.
 *
 * ex_get.c		Command mode input routines.
 *
 * ex_io.c		General input/output processing: file i/o, unix
 *			escapes, filtering, source commands, preserving
 *			and recovering.
 *
 * ex_put.c		Terminal driving and optimizing routines for low-level
 *			output (cursor-positioning); output line formatting
 *			routines.
 *
 * ex_re.c		Global commands, substitute, regular expression
 *			compilation and execution.
 *
 * ex_set.c		The set command.
 *
 * ex_subr.c		Loads of miscellaneous subroutines.
 *
 * ex_temp.c		Editor buffer routines for main buffer and also
 *			for named buffers (Q registers if you will.)
 *
 * ex_tty.c		Terminal dependent initializations from termcap
 *			data base, grabbing of tty modes (at beginning
 *			and after escapes).
 *
 * ex_unix.c		Routines for the ! command and its variations.
 *
 * ex_v*.c		Visual/open mode routines... see ex_v.c for a
 *			guide to the overall organization.
 */

/*
 * This sets the Version of ex/vi for both the exstrings file and
 * the version command (":ver").
 */

/* variable used by ":ver" command */
unsigned char *Version = (unsigned char *)"Version SVR4.0, Solaris 2.5.0";

/*
 * NOTE: when changing the Version number, it must be changed in the
 *	 following files:
 *
 *			  port/READ_ME
 *			  port/ex.c
 *			  port/ex.news
 *
 */
#ifdef XPG4
unsigned char *savepat = (unsigned char *) NULL; /* firstpat storage	*/
#endif /* XPG4 */

/*
 * Main procedure.  Process arguments and then
 * transfer control to the main command processing loop
 * in the routine commands.  We are entered as either "ex", "edit", "vi"
 * or "view" and the distinction is made here. For edit we just diddle options;
 * for vi we actually force an early visual command.
 */

static void usage(unsigned char *);

static int validate_exrc(unsigned char *);

void init(void);

int
main(int ac, char *av[])
{
	extern 	char 	*optarg;
	extern 	int	optind;
	unsigned char	*rcvname = 0;
	unsigned char *cp;
	int c;
	unsigned char	*cmdnam;
	bool recov = 0;
	bool ivis = 0;
	bool itag = 0;
	bool fast = 0;
	extern int verbose;
	int argcounter = 0;
	extern int tags_flag;	/* Set if tag file is not sorted (-S flag) */
	unsigned char scratch [PATH_MAX+1];   /* temp for sourcing rc file(s) */
	int vret = 0;
	unsigned char exrcpath [PATH_MAX+1]; /* temp for sourcing rc file(s) */
	int toptseen = 0;
#ifdef TRACE
	unsigned char *tracef;
#endif
	tagflg = 0;
	(void) setlocale(LC_ALL, "");
#if !defined(TEXT_DOMAIN)
#define	TEXT_DOMAIN "SYS_TEST"
#endif
	(void) textdomain(TEXT_DOMAIN);

	/*
	 * Immediately grab the tty modes so that we won't
	 * get messed up if an interrupt comes in quickly.
	 */
	(void) gTTY(2);
	normf = tty;
	ppid = getpid();
	/* Note - this will core dump if you didn't -DSINGLE in CFLAGS */
	lines = 24;
	columns = 80;	/* until defined right by setupterm */
	/*
	 * Defend against d's, v's, w's, and a's in directories of
	 * path leading to our true name.
	 */
	if ((cmdnam = (unsigned char *)strrchr(av[0], '/')) != 0)
		cmdnam++;
	else
		cmdnam = (unsigned char *)av[0];

	if (EQ((char *)cmdnam, "vi"))
		ivis = 1;
	else if (EQ(cmdnam, "view")) {
		ivis = 1;
		value(vi_READONLY) = 1;
	} else if (EQ(cmdnam, "vedit")) {
		ivis = 1;
		value(vi_NOVICE) = 1;
		value(vi_REPORT) = 1;
		value(vi_MAGIC) = 0;
		value(vi_SHOWMODE) = 1;
	} else if (EQ(cmdnam, "edit")) {
		value(vi_NOVICE) = 1;
		value(vi_REPORT) = 1;
		value(vi_MAGIC) = 0;
		value(vi_SHOWMODE) = 1;
	}

#ifdef	XPG4
	{
		struct winsize jwin;
		char *envptr;

		envlines = envcolumns = -1;
		oldlines = oldcolumns = -1;

		if (ioctl(0, TIOCGWINSZ, &jwin) != -1) {
			oldlines = jwin.ws_row;
			oldcolumns = jwin.ws_col;
		}

		if ((envptr = getenv("LINES")) != NULL &&
		    *envptr != '\0' && isdigit(*envptr)) {
			if ((envlines = atoi(envptr)) <= 0) {
				envlines = -1;
			}
		}

		if ((envptr = getenv("COLUMNS")) != NULL &&
		    *envptr != '\0' && isdigit(*envptr)) {
			if ((envcolumns = atoi(envptr)) <= 0) {
				envcolumns = -1;
			}
		}
	}
#endif /* XPG4 */

	draino();
	pstop();

	/*
	 * Initialize interrupt handling.
	 */
	oldhup = signal(SIGHUP, SIG_IGN);
	if (oldhup == SIG_DFL)
		signal(SIGHUP, onhup);
	oldquit = signal(SIGQUIT, SIG_IGN);
	ruptible = signal(SIGINT, SIG_IGN) == SIG_DFL;
	if (signal(SIGTERM, SIG_IGN) == SIG_DFL)
		signal(SIGTERM, onhup);
	if (signal(SIGEMT, SIG_IGN) == SIG_DFL)
		signal(SIGEMT, onemt);
	signal(SIGILL, oncore);
	signal(SIGTRAP, oncore);
	signal(SIGIOT, oncore);
	signal(SIGFPE, oncore);
	signal(SIGBUS, oncore);
	signal(SIGSEGV, oncore);
	signal(SIGPIPE, oncore);
	init_re();
	while (1) {
#ifdef TRACE
		while ((c = getopt(ac, (char **)av, "VU:Lc:Tvt:rlw:xRCsS")) !=
		    EOF)
#else
		while ((c = getopt(ac, (char **)av,
		    "VLc:vt:rlw:xRCsS")) != EOF)
#endif
			switch (c) {
			case 's':
				hush = 1;
				value(vi_AUTOPRINT) = 0;
				fast++;
				break;

			case 'R':
				value(vi_READONLY) = 1;
				break;
			case 'S':
				tags_flag = 1;
				break;
#ifdef TRACE
			case 'T':
				tracef = (unsigned char *)"trace";
				goto trace;

			case 'U':
				tracef = tttrace;
				strcpy(tracef, optarg);
		trace:
				trace = fopen((char *)tracef, "w");
#define	tracbuf	NULL
				if (trace == NULL)
					viprintf("Trace create error\n");
				else
					setbuf(trace, (char *)tracbuf);
				break;
#endif
			case 'c':
				if (optarg != NULL)
					firstpat = (unsigned char *)optarg;
				else
					firstpat = (unsigned char *)"";
				break;

			case 'l':
				value(vi_LISP) = 1;
				value(vi_SHOWMATCH) = 1;
				break;

			case 'r':
				if (av[optind] && (c = av[optind][0]) &&
				    c != '-') {
					if ((strlen(av[optind])) >=
					    sizeof (savedfile)) {
						(void) fprintf(stderr, gettext(
						    "Recovered file name"
						    " too long\n"));
						exit(1);
					}

					rcvname = (unsigned char *)av[optind];
					optind++;
				}

			case 'L':
				recov++;
				break;

			case 'V':
				verbose = 1;
				break;

			case 't':
				if (toptseen) {
					usage(cmdnam);
					exit(1);
				} else {
					toptseen++;
				}
				itag = tagflg = 1; /* -t option */
				if (strlcpy(lasttag, optarg,
				    sizeof (lasttag)) >= sizeof (lasttag)) {
					(void) fprintf(stderr, gettext("Tag"
					    " file name too long\n"));
					exit(1);
				}
				break;

			case 'w':
				defwind = 0;
				if (optarg[0] == NULL)
					defwind = 3;
				else for (cp = (unsigned char *)optarg;
				    isdigit(*cp); cp++)
					defwind = 10*defwind + *cp - '0';
				if (defwind < 0)
					defwind = 3;
				break;

			case 'C':
			case 'x':
    				(void) fprintf(stderr, gettext("-x and -C "
	    				    "options are no longer "
					    "supported.\n"));
				exit(2);
				break;

			case 'v':
				ivis = 1;
				break;

			default:
				usage(cmdnam);
				exit(1);
			}
		if (av[optind] && av[optind][0] == '+' &&
		    av[optind-1] && strcmp(av[optind-1], "--")) {
				firstpat = (unsigned char *)&av[optind][1];
				optind++;
				continue;
		} else if (av[optind] && av[optind][0] == '-' &&
		    av[optind-1] && strcmp(av[optind-1], "--")) {
			hush = 1;
			value(vi_AUTOPRINT) = 0;
			fast++;
			optind++;
			continue;
		}
		break;
	}

	if (isatty(0) == 0) {
		/*
		 * If -V option is set and input is coming in via
		 * stdin then vi behavior should be ignored. The vi
		 * command should act like ex and only process ex commands
		 * and echo the input ex commands to stderr
		 */
		if (verbose == 1) {
			ivis = 0;
		}

		/*
		 * If the standard input is not a terminal device,
		 * it is as if the -s option has been specified.
		 */
		if (ivis == 0) {
			hush = 1;
			value(vi_AUTOPRINT) = 0;
			fast++;
		}
	}

	ac -= optind;
	av  = &av[optind];

	for (argcounter = 0; argcounter < ac; argcounter++) {
		if ((strlen(av[argcounter])) >= sizeof (savedfile)) {
			(void) fprintf(stderr, gettext("File argument"
			    " too long\n"));
			exit(1);
		}
	}

#ifdef SIGTSTP
	if (!hush && signal(SIGTSTP, SIG_IGN) == SIG_DFL)
		signal(SIGTSTP, onsusp), dosusp++;
#endif


#ifndef PRESUNEUC
	/*
	 * Perform locale-specific initialization
	 */
	localize();
#endif /* PRESUNEUC */

	/*
	 * Initialize end of core pointers.
	 * Normally we avoid breaking back to fendcore after each
	 * file since this can be expensive (much core-core copying).
	 * If your system can scatter load processes you could do
	 * this as ed does, saving a little core, but it will probably
	 * not often make much difference.
	 */
	fendcore = (line *) sbrk(0);
	endcore = fendcore - 2;

	/*
	 * If we are doing a recover and no filename
	 * was given, then execute an exrecover command with
	 * the -r option to type out the list of saved file names.
	 * Otherwise set the remembered file name to the first argument
	 * file name so the "recover" initial command will find it.
	 */
	if (recov) {
		if (ac == 0 && (rcvname == NULL || *rcvname == NULL)) {
			ppid = 0;
			setrupt();
			execlp(EXRECOVER, "exrecover", "-r", (char *)0);
			filioerr(EXRECOVER);
			exit(++errcnt);
		}
		if (rcvname && *rcvname)
			(void) strlcpy(savedfile, rcvname, sizeof (savedfile));
		else {
			(void) strlcpy(savedfile, *av++, sizeof (savedfile));
			ac--;
		}
	}

	/*
	 * Initialize the argument list.
	 */
	argv0 = (unsigned char **)av;
	argc0 = ac;
	args0 = (unsigned char *)av[0];
	erewind();

	/*
	 * Initialize a temporary file (buffer) and
	 * set up terminal environment.  Read user startup commands.
	 */
	if (setexit() == 0) {
		setrupt();
		intty = isatty(0);
		value(vi_PROMPT) = intty;
		if (((cp = (unsigned char *)getenv("SHELL")) != NULL) &&
		    (*cp != '\0')) {
			if (strlen(cp) < sizeof (shell)) {
				(void) strlcpy(shell, cp, sizeof (shell));
			}
		}
		if (fast)
			setterm((unsigned char *)"dumb");
		else {
			gettmode();
			cp = (unsigned char *)getenv("TERM");
			if (cp == NULL || *cp == '\0')
				cp = (unsigned char *)"unknown";
			setterm(cp);
		}
	}

	/*
	 * Bring up some code from init()
	 * This is still done in init later. This
	 * avoids null pointer problems
	 */

	dot = zero = truedol = unddol = dol = fendcore;
	one = zero+1;
	{
	int i;

	for (i = 0; i <= 'z'-'a'+1; i++)
		names[i] = 1;
	}

	if (setexit() == 0 && !fast) {
		if ((globp =
		    (unsigned char *) getenv("EXINIT")) && *globp) {
			if (ivis)
				inexrc = 1;
			commands(1, 1);
			inexrc = 0;
		} else {
			globp = 0;
			if ((cp = (unsigned char *) getenv("HOME")) !=
			    0 && *cp) {
				strncpy(scratch, cp, sizeof (scratch) - 1);
				strncat(scratch, "/.exrc",
				    sizeof (scratch) - 1 - strlen(scratch));
				if (ivis)
					inexrc = 1;
				if ((vret = validate_exrc(scratch)) == 0) {
					source(scratch, 1);
				} else {
					if (vret == -1) {
						error(gettext(
						    "Not owner of .exrc "
						    "or .exrc is group or "
						    "world writable"));
					}
				}
				inexrc = 0;
			}
		}

		/*
		 * Allow local .exrc if the "exrc" option was set. This
		 * loses if . is $HOME, but nobody should notice unless
		 * they do stupid things like putting a version command
		 * in .exrc.
		 * Besides, they should be using EXINIT, not .exrc, right?
		 */

		if (value(vi_EXRC)) {
			if (ivis)
				inexrc = 1;
			if ((cp = (unsigned char *) getenv("PWD")) != 0 &&
			    *cp) {
				strncpy(exrcpath, cp, sizeof (exrcpath) - 1);
				strncat(exrcpath, "/.exrc",
				    sizeof (exrcpath) - 1 - strlen(exrcpath));
				if (strcmp(scratch, exrcpath) != 0) {
					if ((vret =
					    validate_exrc(exrcpath)) == 0) {
						source(exrcpath, 1);
					} else {
						if (vret == -1) {
							error(gettext(
							    "Not owner of "
							    ".exrc or .exrc "
							    "is group or world "
							    "writable"));
						}
					}
				}
			}
			inexrc = 0;
		}
	}

	init();	/* moved after prev 2 chunks to fix directory option */

	/*
	 * Initial processing.  Handle tag, recover, and file argument
	 * implied next commands.  If going in as 'vi', then don't do
	 * anything, just set initev so we will do it later (from within
	 * visual).
	 */
	if (setexit() == 0) {
		if (recov)
			globp = (unsigned char *)"recover";
		else if (itag) {
			globp = ivis ? (unsigned char *)"tag" :
			    (unsigned char *)"tag|p";
#ifdef	XPG4
			if (firstpat != NULL) {
				/*
				 * if the user specified the -t and -c
				 * flags together, then we service these
				 * commands here. -t is handled first.
				 */
				savepat = firstpat;
				firstpat = NULL;
				inglobal = 1;
				commands(1, 1);

				/* now handle the -c argument: */
				globp = savepat;
				commands(1, 1);
				inglobal = 0;
				globp = savepat = NULL;

				/* the above isn't sufficient for ex mode: */
				if (!ivis) {
					setdot();
					nonzero();
					plines(addr1, addr2, 1);
				}
			}
#endif /* XPG4 */
		} else if (argc)
			globp = (unsigned char *)"next";
		if (ivis)
			initev = globp;
		else if (globp) {
			inglobal = 1;
			commands(1, 1);
			inglobal = 0;
		}
	}

	/*
	 * Vi command... go into visual.
	 */
	if (ivis) {
		/*
		 * Don't have to be upward compatible
		 * by starting editing at line $.
		 */
#ifdef	XPG4
		if (!itag && (dol > zero))
#else /* XPG4 */
		if (dol > zero)
#endif /* XPG4 */
			dot = one;
		globp = (unsigned char *)"visual";
		if (setexit() == 0)
			commands(1, 1);
	}

	/*
	 * Clear out trash in state accumulated by startup,
	 * and then do the main command loop for a normal edit.
	 * If you quit out of a 'vi' command by doing Q or ^\,
	 * you also fall through to here.
	 */
	seenprompt = 1;
	ungetchar(0);
	globp = 0;
	initev = 0;
	setlastchar('\n');
	setexit();
	commands(0, 0);
	cleanup(1);
	return (errcnt);
}

/*
 * Initialization, before editing a new file.
 * Main thing here is to get a new buffer (in fileinit),
 * rest is peripheral state resetting.
 */
void
init(void)
{
	int i;
	fileinit();
	dot = zero = truedol = unddol = dol = fendcore;
	one = zero+1;
	undkind = UNDNONE;
	chng = 0;
	edited = 0;
	for (i = 0; i <= 'z'-'a'+1; i++)
		names[i] = 1;
	anymarks = 0;
}

/*
 * Return last component of unix path name p.
 */
unsigned char *
tailpath(p)
unsigned char *p;
{
	unsigned char *r;

	for (r = p; *p; p++)
		if (*p == '/')
			r = p+1;
	return (r);
}


/*
 * validate_exrc - verify .exrc as belonging to the user.
 * The file uid should match the process ruid,
 * and the file should be writable only by the owner.
 */
static int
validate_exrc(unsigned char *exrc_path)
{
	struct stat64 exrc_stat;
	int process_uid;

	if (stat64((char *)exrc_path, &exrc_stat) == -1)
		return (0); /* ignore if .exrec is not found */
	process_uid = geteuid();
	/* if not root, uid must match file owner */
	if (process_uid && process_uid != exrc_stat.st_uid)
		return (-1);
	if ((exrc_stat.st_mode & (S_IWGRP | S_IWOTH)) != 0)
		return (-1);
	return (0);
}

/*
 * print usage message to stdout
 */
static void
usage(unsigned char *name)
{
	char buf[160];

#ifdef TRACE
	(void) snprintf(buf, sizeof (buf), gettext(
	    "Usage: %s [- | -s] [-l] [-L] [-wn] "
	    "[-R] [-S] [-r [file]] [-t tag] [-T] [-U tracefile]\n"
	    "[-v] [-V] [-x] [-C] [+cmd | -c cmd] file...\n"), name);
#else
	(void) snprintf(buf, sizeof (buf), gettext(
	    "Usage: %s [- | -s] [-l] [-L] [-wn] "
	    "[-R] [-S] [-r [file]] [-t tag]\n"
	    "[-v] [-V] [-x] [-C] [+cmd | -c cmd] file...\n"), name);
#endif
	(void) write(2, buf, strlen(buf));
}
